### PR TITLE
[DEV-3911] logical repo fix clean deployments

### PR DESCRIPTION
**Pull request by gaetano-guerriero**
While inspecting code for DEV-3911 I spotted a problem on logical deployments cleaning.

Also a small refactor has been done to reduce code duplication.

### CODE REVIEW

File: server/athenian/api/controllers/settings_controller.py [hunk 1]
```diff
@@ -462,9 +462,15 @@ async def set_logical_repository(request: AthenianWebRequest, body: dict) -> web
                             web_model.name, full_name, prefixed_name, repo_id,
                             web_model.account, sdb_conn, request.pdb)
                     break
+            else:
+                pass
+                # TODO: what happens here ?? no deletion and there's a following insert
```
Comment (gaetano-guerriero) [inline server/athenian/api/controllers/settings_controller.py:467]: other open issue

File: server/athenian/api/controllers/settings_controller.py [hunk 2]
```diff
@@ -566,26 +545,38 @@ async def clean_repository_sets():
             model.acc_id == account,
             model.repository_full_name == full_name,
         ))))
+    clean_deployments_tasks = await _clean_logical_deployments_tasks(
+        [full_name, drop_logical_repo(full_name)], account, pdb,
+    )
+    await gather(*tasks, *clean_deployments_tasks, op="_delete_logical_repository/sql")
```
Comment (gaetano-guerriero) [inline server/athenian/api/controllers/settings_controller.py:551]: Personally I would make the code simpler by making `_clean_logical_deployments..` function execute all queries.
We would then lose a bit parallelism here but I maybe response time is not important in these 2 endpoints

File: server/tests/controllers/test_settings_controller.py [hunk 3]
```diff
@@ -1022,6 +1092,72 @@ async def test_set_logical_repository_replace(
     await _test_set_logical_repository(client, headers, sdb, 2)
 
 
+# TODO: replacing existing identical logical repository is not supported
+@pytest.mark.xfail
```
Comment (gaetano-guerriero) [inline server/tests/controllers/test_settings_controller.py:1096]: this is the issue for the # TODO added

File: server/athenian/api/controllers/settings_controller.py [hunk 4]
```diff
@@ -566,26 +544,39 @@ async def clean_repository_sets():
...
+    await gather(*tasks, clean_deployments_coro, op="_delete_logical_repository/sql")
+
+
+async def _clean_logical_deployments(
+    repos: Sequence[str], account: int, pdb: DatabaseLike, op: str = None,
+) -> None:
+    """Delete all deployments info about `repos` from pdb."""
+    depl_models = (GitHubCommitDeployment, GitHubPullRequestDeployment, GitHubReleaseDeployment)
     affected_deployments = await pdb.fetch_all(union(*(
         select([distinct(model.deployment_name)]).where(and_(
```
Comment (gaetano-guerriero) [inline server/athenian/api/controllers/settings_controller.py:559]: `delete ... returning` not yet in sqlalchemy with sqlite engine

File: server/athenian/api/controllers/settings_controller.py [hunk 5]
```diff
@@ -566,26 +544,39 @@ async def clean_repository_sets():
...
+        [full_name, drop_logical_repo(full_name)], account, pdb,
+    )
+    await gather(*tasks, clean_deployments_coro, op="_delete_logical_repository/sql")
+
+
+async def _clean_logical_deployments(
+    repos: Sequence[str], account: int, pdb: DatabaseLike, op: str = None,
+) -> None:
+    """Delete all deployments info about `repos` from pdb."""
+    depl_models = (GitHubCommitDeployment, GitHubPullRequestDeployment, GitHubReleaseDeployment)
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/settings_controller.py:557]:
```suggestion
    deployment_models = (GitHubCommitDeployment, GitHubPullRequestDeployment, GitHubReleaseDeployment)
```
(`depl` is uncommon)

File: server/athenian/api/controllers/settings_controller.py [hunk 6]
```diff
@@ -566,26 +544,39 @@ async def clean_repository_sets():
...
         ))
-        for model in (GitHubCommitDeployment, GitHubPullRequestDeployment, GitHubReleaseDeployment)
+        for model in depl_models
     )))
+    tasks = []
     if affected_deployments:
-        for model in (GitHubCommitDeployment,
-                      GitHubPullRequestDeployment,
-                      GitHubReleaseDeployment):
-            tasks.append(delete(model).where(and_(
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/settings_controller.py:580]: :man_facepalming:  I guess this code was not tested at all

File: server/athenian/api/controllers/settings_controller.py (hunk 4 above)
Comment (vmarkovtsev) [reply server/athenian/api/controllers/settings_controller.py:559]: Yes. BTW can you please `log.info()` the `[r[0] for r in affected_deployments]`, it should simplify our life in the future.

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]
Eagle-eyed code reading :eagle:

File: server/athenian/api/controllers/settings_controller.py (hunk 1 above)
Comment (gaetano-guerriero) [reply server/athenian/api/controllers/settings_controller.py]: done, please check if release settings match comparison is right

File: server/athenian/api/controllers/settings_controller.py (hunk 4 above)
Comment (gaetano-guerriero) [reply server/athenian/api/controllers/settings_controller.py:596]: added

File: server/athenian/api/controllers/settings_controller.py [hunk 7]
```diff
@@ -493,35 +498,43 @@ async def set_logical_repository(request: AthenianWebRequest, body: dict) -> web
...
-                                     account: int,
-                                     pdb: DatabaseLike) -> None:
-    affected_deployments = await pdb.fetch_all(union(*(
-        select([distinct(model.deployment_name)]).where(and_(
-            model.acc_id == account,
-            model.repository_full_name == repo,
-        ))
-        for model in (GitHubCommitDeployment, GitHubPullRequestDeployment, GitHubReleaseDeployment)
+async def _find_matching_logical_repository(
+    logical_repo: LogicalRepository,
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/settings_controller.py:505]: `logical_repo` duplicates `logical_repo_request` in almost everything, not good. How about consuming `parent_name` and `ReleaseSettings` instead?

File: server/athenian/api/controllers/settings_controller.py [hunk 8]
```diff
@@ -566,26 +579,45 @@ async def clean_repository_sets():
...
-                model.repository_full_name.in_([full_name, drop_logical_repo(full_name)]),
-            )))
-        tasks.append(delete(GitHubDeploymentFacts).where(and_(
-            GitHubDeploymentFacts.acc_id == account,
-            GitHubDeploymentFacts.deployment_name.in_(affected_deployments),
-        )))
-    await gather(*tasks, op="_delete_logical_repository/sql")
+        log = logging.getLogger(f"{metadata.__package__}._clean_logical_deployments")
+        log.info(
+            "Cleaning deployments %s for repos %s", [r[0] for r in affected_deployments], repos,
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/settings_controller.py:606]: We should save `[r[0] for r in affected_deployments]` to a variable, we use it two times.

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: server/athenian/api/controllers/settings_controller.py (hunk 1 above)
Comment (vmarkovtsev) [reply server/athenian/api/controllers/settings_controller.py]: It's fine :+1:

File: server/athenian/api/controllers/settings_controller.py (hunk 7 above)
Comment (gaetano-guerriero) [reply server/athenian/api/controllers/settings_controller.py:507]: done

File: server/athenian/api/controllers/settings_controller.py (hunk 8 above)
Comment (gaetano-guerriero) [reply server/athenian/api/controllers/settings_controller.py]: done

**Review by vmarkovtsev — APPROVED** [review]

**Pull request merged**